### PR TITLE
APIs needed for first prod release

### PR DESCRIPTION
**Pull request by webb04**
Functions should allow us to:
- Follow authors
- Unfollow authors
- Hide/Show ads
- Insert epics (same API as in the current templates)

### CODE REVIEW

File: thrift/native.thrift [hunk 1]
```diff
@@ -8,4 +8,7 @@ struct AdSlot {
 service Native {
     void insertAdverts(1:list<AdSlot> adSlots),
     i32 nativeThriftPackageVersion(),
+    void followAuthor(),
+    void unfollowAuthor(),
```
Comment (JamieB-gu) [inline thrift/native.thrift]: Do these not need to take an author identifier of some sort? Or does the native layer already know?

File: thrift/webview.thrift [hunk 2]
```diff
@@ -1,3 +1,10 @@
+struct Epic {
+    1: required string title;
+    2: required string body;
+    3: required string firstButton;
+    4: optional string secondButton;
```
Comment (JamieB-gu) [inline thrift/webview.thrift]: Will any of these strings contain html?

File: thrift/native.thrift (hunk 1 above)
Comment (webb04) [reply thrift/native.thrift]: They should know by the context of the open article, but we can pass back an ID if it's easier for them. We can also have one function to toggle following too.

File: thrift/webview.thrift (hunk 2 above)
Comment (webb04) [reply thrift/webview.thrift]: Yep the first two I think

**Comment by webb04** [comment]
Realised we'll need another function for registering epic button taps.... Or an api for opening the friction screen.

**Comment by davidfurey** [comment]
> Do we have an view of how @alexduf's idea of a `ready` event might work within this framework? Do we want to send one general `ready` event that encompasses things like ad insertion, or do we want to signal each startup event through individual functions?

The `ready` event is just to tell the native layer that the webview thrift API is available to be called.  I think that should be independent of the JS asking the native layer to place some ads on top of the page.

File: thrift/native.thrift (hunk 1 above)
Comment (alexduf) [reply thrift/native.thrift]: I think it should be `followTopic(topic: Topic)`, as you can follow contributors, matches or liveblogs

**Comment by alexduf** [comment]
I'm still torn between having the webview being the source of all the events (ophan, permutive, ads etc) or giving more context to the app and letting the app deal with it. I think we probably want to chat about it next time we meet regarding the SSR project

File: thrift/native.thrift (hunk 1 above)
Comment (alexanderedge) [reply thrift/native.thrift]: Do we also need `isFollowingTopic(1: string identifier)` so that the webview can know if it's already following something?

Comment (alexanderedge) [reply thrift/native.thrift]: Should follow / unfollow be asynchronous? If the action fails, then the webview may show an incorrect state.

Comment (maxspencer) [reply thrift/native.thrift]: > I think it should be `followTopic(topic: Topic)`, as you can follow contributors, matches or liveblogs

There could be multiple "Follow X", "Follow Y" buttons on the same article too, so identifying X and Y is necessary.

Comment (webb04) [reply thrift/native.thrift]: What would these look like? is it a contributor id?

Comment (davidfurey) [reply thrift/native.thrift]: Since we have types, could it not just be `follow(topic: Topic)`?

Comment (webb04) [reply thrift/native.thrift]: What would `Topic` look like? what would be best for native layers?

Comment (alexanderedge) [reply thrift/native.thrift]: On iOS, Topic is a structure with a type (which is a [string enum](https://github.com/guardian/ios-live/blob/ad7a86535f1a654dea21ad95551a0450bd60daf4/GLA/GLA/Classes/MAPIEnums.m#L38)) and a name. An identifier is created from the type and the name by concatenating in the format `{type}///{name}`.

Comment (alexanderedge) [reply thrift/native.thrift]: Just a thought: it might be worth creating a separate PR for each function that we add

Comment (webb04) [reply thrift/native.thrift]: Makes sense so each function will be a new version. We can chat today, but it looks like we'll need these and `isFollowing` and `updateAdPosition`

**Comment by webb04** [comment]
Opened up some smaller PRs:
- https://github.com/guardian/mobile-apps-thrift/pull/5
- https://github.com/guardian/mobile-apps-thrift/pull/6
- https://github.com/guardian/mobile-apps-thrift/pull/7

**Pull request closed without merging**